algorithms/src/std_algorithms/Kokkos_PartitioningOperations.hpp

```diff
@@ -0,0 +1,489 @@
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_PARTITIONING_OPERATIONS_HPP
+#define KOKKOS_STD_PARTITIONING_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+#include "Kokkos_MinMaxOperations.hpp"
+#include "Kokkos_ModifyingOperations.hpp"
+#include "Kokkos_NonModifyingSequenceOperations.hpp"
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// -------------------------
+//
+// functors
+//
+// -------------------------
+
+template <class IteratorType, class ReducerType, class PredicateType>
+struct StdIsPartitionedFunctor {
+  using red_value_type = typename ReducerType::value_type;
+  using index_type     = typename IteratorType::difference_type;
+
+  IteratorType m_first;
+  ReducerType m_reducer;
+  PredicateType m_p;
+
+  KOKKOS_FUNCTION
+  void operator()(const index_type i, red_value_type& redValue) const {
+    const auto predicate_value = m_p(m_first[i]);
+    constexpr index_type m_red_id_min =
+        ::Kokkos::reduction_identity<index_type>::min();
+    constexpr index_type m_red_id_max =
+        ::Kokkos::reduction_identity<index_type>::max();
+    auto rv = predicate_value ? red_value_type{i, m_red_id_min}
+                              : red_value_type{m_red_id_max, i};
+
+    m_reducer.join(redValue, rv);
+  }
+
+  KOKKOS_FUNCTION
+  StdIsPartitionedFunctor(IteratorType first, ReducerType reducer,
+                          PredicateType p)
+      : m_first(first), m_reducer(std::move(reducer)), m_p(std::move(p)) {}
+};
+
+template <class IteratorType, class ReducerType, class PredicateType>
+struct StdPartitionPointFunctor {
+  using red_value_type = typename ReducerType::value_type;
+  using index_type     = typename IteratorType::difference_type;
+
+  IteratorType m_first;
+  ReducerType m_reducer;
+  PredicateType m_p;
+
+  KOKKOS_FUNCTION
+  void operator()(const index_type i, red_value_type& redValue) const {
+    const auto predicate_value = m_p(m_first[i]);
+    auto rv =
+        predicate_value
+            ? red_value_type{::Kokkos::reduction_identity<index_type>::min()}
+            : red_value_type{i};
+    m_reducer.join(redValue, rv);
+  }
+
+  KOKKOS_FUNCTION
+  StdPartitionPointFunctor(IteratorType first, ReducerType reducer,
+                           PredicateType p)
+      : m_first(first), m_reducer(std::move(reducer)), m_p(std::move(p)) {}
+};
+
+template <class ValueType>
+struct StdPartitionCopyScalar {
+  ValueType true_count_;
+  ValueType false_count_;
+
+  KOKKOS_FUNCTION
+  StdPartitionCopyScalar& operator=(const StdPartitionCopyScalar& other) {
+    true_count_  = other.true_count_;
+    false_count_ = other.false_count_;
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  void operator=(const volatile StdPartitionCopyScalar& other) {
+    true_count_  = other.true_count_;
+    false_count_ = other.false_count_;
+  }
```

Comment:
I did that for consistency with all the "scalar" structs are defined for the reductions in `Kokkos_Parallel_Reduce.hpp`. 
For example: 
```cpp
template <class Scalar>
struct MinMaxScalar {
  Scalar min_val, max_val;

  KOKKOS_INLINE_FUNCTION
  void operator=(const MinMaxScalar& rhs) {
    min_val = rhs.min_val;
    max_val = rhs.max_val;
  }

  KOKKOS_INLINE_FUNCTION
  void operator=(const volatile MinMaxScalar& rhs) volatile {
    min_val = rhs.min_val;
    max_val = rhs.max_val;
  }
};
```

(btw, I just noticed I am missing a volatile in line 131 to qualify the method, so I will add that.)

Reply:
I added a comment on this in the code

Reply:
I would very much prefer if we `default` the regular copy assignment operator but we can do that in another pull request consistently...

Reply:
Ok so i leave it for now and we can change that later. I also think that defaulting might cause some compiler error for the volatile overloads when dealing with openmptarget and similar. Because of the fact that these operators should return "type &". I think that is also why these overloads are explicitly implemented as returning void. I remember seeing something like that before.

Reply:
Yeah, I was seeing that we want the `volatile` overload to return `void` to suppress some annoying compiler warnings but the regular one should be fine returning `type&` (and we need to define it somehow, `= default` should be enough, to avoid using the `volatile` overload).

Reply:
i am marking this as resolved since we agree on addressing that in another PR ok?

Reply:
Sure if you don't forget about it. :slightly_smiling_face:

Reply:
I opened this https://github.com/kokkos/kokkos/issues/4346 so we don't forget :)